### ISSUE TITLE
Can Bull be Used in a Remote Polling Scenario?

### ISSUE THREAD

**Issue by travissutton** [OptimalBits/bull#790]
I'm currently leveraging bull in our project for it's seemingly intended purpose but also have the need to implement a queue system that allows for remote workers to poll an API for tasks to be completed. From what I'm reading on the documentation it doesn't look like this is going to work as the remote workers have no connection to the API's redis connection. Rather than doing a `queue.process` I need to do something like `queue.poll` to get a job, then have that job locked for some duration. The worker would then post back to the API with updates, the API would then need to be able to readily access the Job on-demand (assuming `queue.getJob`). Finally, I would need to either complete or fail the job from the API side based on a final update from the worker (all I can find here is `queue.getJob`->`job.progress(100)` - hoping this actually closes out a job??).

I'm going to try this with `queue.getWaiting` but from the documentation, this doesn't seem like I'm going to be able to still do what I ultimately am trying to do.

Is remote polling/job work-flow not supported in bull currently? If so, are there API calls that are not documented that I should be using? Thanks for the help.

**Comment by manast** [comment]
ok. Not sure I completely understand your case, but If I got it right what you want is to be able to process jobs on demand instead of processing them as soon as they arrive the queue, is it correct?
There is no API for doing that, but you probably could write such a function yourself and monkey patch the queue (or submit a PR if it makes sense to have it as a feature in bull). Basically you should write a method that does a ```getNextJob```( https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L920) and directly after a ```processJob```(https://github.com/OptimalBits/bull/blob/master/lib/queue.js#L838) where you specify your handler function (processing function for the job) and you probably do not want to call blpoprpush and instead just call ```moveToActive``` which returns a job if there is any or undefined otherwise. Good luck!

**Comment by travissutton** [comment]
You have it correct for the most part. I have it working using private functions to see if it was even possible. Here's the workflow:

1) Job comes into queue via some other service.
2) Remote agent asks queue service over http if there are any jobs via `queue.getNextJob`, so that it get the `moveToActive` call made.
3) Queue service returns `Job` data to agent, agent parses data and does work. This includes job.id for referencing back to Queue Service
4) Agent provides update to Queue Service via http - progress, etc - using combination of `queue.getJob` and `job.progress`
5) Agent provides completed notification to Queue service via http - Queue service does `queue.getJob`->`job.progress(100)`->`job.moveToCompleted`
6) Agent asks Queue Service if it has another task to do (restart)

The requirement of being able to process on-demand was the key. The end result seems like it may work really well, though I don't like that I'm using private functions. Are there any plans to make those methods public? Specifically the `getNextJob` method?

**Comment by manast** [comment]
No, we do not have plans for that I am afraid. I would recommend that you write a few unit tests specifically for testing the code using the private functions, in that way hopefully you will catch eventual changes to those APIs when upgrading and you will have a chance to react accordingly.

**Comment by aleccool213** [comment]
@travissutton Got anywhere with this? Me and my team are trying to do with the exact same thing.

**Comment by aleccool213** [comment]
I am going to make a PR which specifies this as a pattern and possibly exposes some private funcs.
